File: client/src/main/java/pro/panopticon/client/sensor/impl/MemorySensor.java
package pro.panopticon.client.sensor.impl;

import com.amazonaws.services.cloudwatch.model.StandardUnit;
import pro.panopticon.client.model.Measurement;
import pro.panopticon.client.model.MetricDimension;
import pro.panopticon.client.sensor.Sensor;
import pro.panopticon.client.util.SystemStatus;

import java.util.*;

public class MemorySensor implements Sensor {
    private static final String DESCRIPTION = "When this alarm is triggered, you should check the memory status of the other nodes as well. " +
            "There might be a memory leak somewhere in the application triggering this, so a restart will buy you some time";
    private static final long BYTES_IN_MB = 1024 * 1024;

    private final int warnLimitNow;
    private final int errorLimitNow;
    private final int warnLimitHeap;
    private final int errorLimitHeap;
    private final Optional<String> hostname;

    public MemorySensor() {
        this(85, 95, 75, 95);
    }

    public MemorySensor(int warnLimitNow, int errorLimitNow, int warnLimitHeap, int errorLimitHeap) {
        this(warnLimitNow, errorLimitNow, warnLimitHeap, errorLimitHeap, null);
    }

    public MemorySensor(int warnLimitNow, int errorLimitNow, int warnLimitHeap, int errorLimitHeap, String hostname) {
        this.warnLimitNow = warnLimitNow;
        this.errorLimitNow = errorLimitNow;
        this.warnLimitHeap = warnLimitHeap;
        this.errorLimitHeap = errorLimitHeap;
        this.hostname = Optional.ofNullable(hostname);
    }

    @Override
    public List<Measurement> measure() {
        SystemStatus s = new SystemStatus();

        List<Measurement> measurements = new ArrayList<>();

        putMemoryStatus(measurements, "mem.heap.now", s.heapUsed(), s.heapMax(), warnLimitNow, errorLimitNow);
        putMemoryStatus(measurements, "mem.heap.lastGC", s.heapAfterGC(), s.heapMax(), warnLimitHeap, errorLimitHeap);

        return measurements;
    }

    private void putMemoryStatus(List<Measurement> measurements, String key, long used, long max, int warnLimit, int errorLimit) {

        if (max == 0 || used == -1) {
            return;
        }
        long percentUsed = used / (max / 100);

        String displayValue = toMB(used) + " of " + toMB(max) + " MB (" + percentUsed + "%)";

        List<MetricDimension> dimensions = hostname
                .map(h -> Collections.singletonList(MetricDimension.instanceDimension(h)))
                .orElse(Collections.emptyList());

        measurements.add(
                new Measurement(
                        key,
                        status(percentUsed, warnLimit, errorLimit),
                        displayValue,
                        new Measurement.CloudwatchValue(
                                percentUsed,
                                StandardUnit.Percent,
                                dimensions
                                ),
                        DESCRIPTION));
    }

    private String status(long percentUsed, int warnLimit, int errorLimit) {
        if (percentUsed > errorLimit) {
            return "ERROR";
        } else if (percentUsed > warnLimit) {
            return "WARN";
        } else {
            return "INFO";
        }
    }

    private long toMB(final long bytes) {
        return bytes / BYTES_IN_MB;
    }
}

File: client/src/main/java/pro/panopticon/client/sensor/impl/MemorySensor.kt
package pro.panopticon.client.sensor.impl

import com.amazonaws.services.cloudwatch.model.StandardUnit
import pro.panopticon.client.model.Measurement
import pro.panopticon.client.model.MetricDimension.Companion.instanceDimension
import pro.panopticon.client.sensor.Sensor
import pro.panopticon.client.util.SystemStatus

class MemorySensor @JvmOverloads constructor(
    private val warnLimitNow: Int = 85,
    private val errorLimitNow: Int = 95,
    private val warnLimitHeap: Int = 75,
    private val errorLimitHeap: Int = 95,
    private val hostname: String? = null,
) : Sensor {

    override fun measure(): List<Measurement> {
        return SystemStatus().let {
            listOfNotNull(
                createMeasurement("mem.heap.now", it.heapUsed(), it.heapMax(), warnLimitNow, errorLimitNow),
                createMeasurement("mem.heap.lastGC", it.heapAfterGC(), it.heapMax(), warnLimitHeap, errorLimitHeap)
            )
        }
    }

    private fun createMeasurement(
        key: String,
        used: Long,
        max: Long,
        warnLimit: Int,
        errorLimit: Int,
    ): Measurement? {
        if (max == 0L || used == -1L) {
            return null
        }
        val percentUsed = used / (max / 100)
        val displayValue = toMB(used).toString() + " of " + toMB(max) + " MB (" + percentUsed + "%)"
        val dimensions = hostname?.let { instanceDimension(it) }
                             ?.let { listOf(it) }
                         ?: emptyList()

        return Measurement(
            key = key,
            status = status(percentUsed, warnLimit, errorLimit),
            displayValue = displayValue,
            cloudwatchValue = Measurement.CloudwatchValue(
                percentUsed.toDouble(),
                StandardUnit.Percent,
                dimensions,
            ),
            description = DESCRIPTION,
        )
    }

    private fun status(percentUsed: Long, warnLimit: Int, errorLimit: Int): String {
        return when {
            percentUsed > errorLimit -> "ERROR"
            percentUsed > warnLimit -> "WARN"
            else -> "INFO"
        }
    }

    private fun toMB(bytes: Long): Long {
        return bytes / BYTES_IN_MB
    }

    companion object {
        private const val DESCRIPTION =
            "When this alarm is triggered, you should check the memory status of the other nodes as well. " +
            "There might be a memory leak somewhere in the application triggering this, so a restart will buy you some time"
        private const val BYTES_IN_MB = (1024 * 1024).toLong()
    }
}
